### ISSUE TITLE
How to write to buffer with `avif` format?

### ISSUE THREAD

**Issue by monochromer** [jcupitt/lua-vips#89]
I'm trying convert `png` image to `avif` buffer:

```lua
local vips = require('vips')

local image = vips.Image.new_from_file('image.png')
local buffer = image:write_to_buffer('.avif', { Q = 75, compression = 'av1' })

print('buffer, ', buffer)
```

But get the error: `VipsForeignSave: ".avif" is not a known buffer format`
```
luajit: /usr/local/share/lua/5.1/vips/Image_methods.lua:391: VipsForeignSave: ".avif" is not a known buffer format

stack traceback:
  [C]: in function 'error'
  /usr/local/share/lua/5.1/vips/Image_methods.lua:391: in function 'write_to_buffer'
  ./image.lua:16: in main chunk
  [C]: at 0x55e901d0ced0
```

Method of saving to `avif` file is working successfully:

```lua
image:write_to_file('image.avif', {
   Q = 75,
   effort = 6
})
```

<details>
<summary>my docker environment</summary>

```Dockerfile
FROM alpine:3.20.3

RUN apk add --no-cache --update \
  vips \
  vips-dev \
  vips-heif \
  libheif \
  libheif-dev \
  libheif-pixbuf-loader \
  libheif-tools \
  vips-tools

RUN apk add --no-cache --update \
  lua5.1 \
  luajit \
  luarocks \
  luarocks5.1

RUN luarocks-5.1 config --lua-version=5.1 rocks_provided.luaffi-tkl "2.1-1"
RUN luarocks-5.1 install lua-vips
```

</details>

**Comment by RiskoZS** [comment]
Try passing in `'.heif'` instead of `'.avif'`, since [it seems that's what libvips calls it](https://www.libvips.org/API/current/VipsForeignSave.html#vips-heifsave-buffer). With that, the code snippet you provided gives me an AVIF buffer as expected.

**Comment by monochromer** [comment]
@RiskoZoSlovenska thanks!

**Comment by RiskoZS** [comment]
No worries.

Though it is a bit curious that libvips understands the `.avif` extension when saving to a file but not when saving to a buffer; for example, passing in `'.heic'` works as expected in both scenarios. Unfortunately, I don't know enough about libvips or the details of this binding to tell you why works the way it does or whether this behaviour is even intended.

**Comment by rolandlo** [comment]
> No worries.
> 
> Though it is a bit curious that libvips understands the `.avif` extension when saving to a file but not when saving to a buffer; for example, passing in `'.heic'` works as expected in both scenarios. Unfortunately, I don't know enough about libvips or the details of this binding to tell you why works the way it does or whether this behaviour is even intended.

The binding just calls `vips_foreign_find_save` and `vips_foreign_find_save_buffer`, respectively, from libvips. There is a difference in `vips_foreign_find_save_sub` and `vips_foreign_find_save_buffer_sub`: The latter takes only foreign savers, whose object class nicknames end with "_buffer". So ultimately it comes down to `vips_foreign_save_heif_buffer_class_init` in `heifsave.c` where 
```c
foreign_class->suffs = vips__heic_suffs
```
with
```c
const char *vips__heic_suffs[] = {
	".heic",
	".heif",
	NULL
};

const char *vips__avif_suffs[] = {
	".avif",
	NULL
};

const char *vips__heif_suffs[] = { 
	".heic",
	".heif",
	".avif",
	NULL 
};
```
is defined.

I can't tell why it's not
```c
foreign_class->suffs = vips__heif_suffs
```
@jcupitt can you tell?

**Comment by jcupitt** [comment]
Hello @rolandlo, it's a problem with the design of the saver API -- by the time `vips_foreign_save_heif_buffer_build()` runs, the suffix that the user passed in has been discarded, so it wouldn't know to write as AVIF.

The new target API is supposed to fix this. In python (for example) you can write:

```python
#!/usr/bin/env python3
  
import sys
import pyvips

infilename = sys.argv[1]
fmt = sys.argv[2]

image = pyvips.Image.new_from_file(infilename)

target = pyvips.Target.new_to_descriptor(sys.stdout.fileno())
image.write_to_target(target, fmt)
```

That's using a `new_to_descriptor()` target, but you can have memory, file, pipe, descriptor and custom targets too. 

You can run that code like this:

```
$ ./target.py ~/pics/k2.png .avif > x
$ file x
x: ISO Media, AVIF Image
```

So it's automatically making an AVIF, as expected.

tldr: The older `_buffer` API has been more or less deprecated for a while and we have the source / target generic IO system instead. The new system should be quite a bit nicer, and fixes a whole range of problems like this.

There was a post on libvips.org introducing the new system:

https://www.libvips.org/2019/11/29/True-streaming-for-libvips.html

**Comment by jcupitt** [comment]
The source / target API isn't yet wrapped for lua-vips, I should have said. It's not a huge amount of work, but no one's done it. The ruby and python sources for these classes are here, for reference, they might be useful:

https://github.com/libvips/pyvips/blob/master/pyvips/vconnection.py
https://github.com/libvips/pyvips/blob/master/pyvips/vsource.py
https://github.com/libvips/pyvips/blob/master/pyvips/vsourcecustom.py
https://github.com/libvips/pyvips/blob/master/pyvips/vtarget.py
https://github.com/libvips/pyvips/blob/master/pyvips/vtargetcustom.py

https://github.com/libvips/ruby-vips/blob/master/lib/vips/connection.rb
etc.

The `custom` classes are for connecting your own read / write methods with callbacks and are slightly tricker, but perhaps not strictly necessary.

**Comment by rolandlo** [comment]
> The source / target API isn't yet wrapped for lua-vips, I should have said. It's not a huge amount of work, but no one's done it. The ruby and python sources for these classes are here, for reference, they might be useful:
> 
> https://github.com/libvips/pyvips/blob/master/pyvips/vconnection.py https://github.com/libvips/pyvips/blob/master/pyvips/vsource.py https://github.com/libvips/pyvips/blob/master/pyvips/vsourcecustom.py https://github.com/libvips/pyvips/blob/master/pyvips/vtarget.py https://github.com/libvips/pyvips/blob/master/pyvips/vtargetcustom.py
> 
> https://github.com/libvips/ruby-vips/blob/master/lib/vips/connection.rb etc.
> 
> The `custom` classes are for connecting your own read / write methods with callbacks and are slightly tricker, but perhaps not strictly necessary.

Thanks for the links. I will take a look.

**Comment by monochromer** [comment]
Thanks for the clarification

**Comment by rolandlo** [comment]
> The source / target API isn't yet wrapped for lua-vips, I should have said. It's not a huge amount of work, but no one's done it. 

I'm working on it. See https://github.com/rolandlo/lua-vips/tree/add-true-streaming
